website/src/middleware.ts

```diff
@@ -14,5 +16,38 @@ export const config = {
     "/tasks/:path*",
     "/leaderboard",
     "/messages/:path*",
+    "/api/auth/signin/email",
   ],
 };
+
+const middleware = async (req: NextRequestWithAuth) => {
+  if (req.method === "POST" && req.nextUrl.pathname === "/api/auth/signin/email") {
+    const data = await getBody(req);
+    const res = await checkCaptcha(data?.captcha, req.ip);
+
+    if (res.success) {
+      return NextResponse.next();
+    }
+
+    const url = req.nextUrl.clone();
+    url.pathname = "/api/auth/invalid-captcha";
+
+    return NextResponse.redirect(url);
+  }
+
+  return withAuth(req);
+};
```

Comment:
This is extremely hacky. I tried 2 other solutions but they did not work.
- `NextResponse.redirect('/auth/singin?errorCode=somecode')` cause next-auth `signIn` function doesn't follow redirect.
- `return NextResponse.json({url: '/auth/singin?errorCode=somecode'})` cause Nextjs doesn't allow to modify response body in middleware

Reply:
did you try using the signIn callback within NextAuth? If you modify `src/pages/api/auth/[...[nextauth].ts` You can fill out the `signIn` call back.  Returning true means signin works.  Returning false rejects the signin.  So I think you could put very similar code in that callback when the user is using the email provider.  Not entirely sure how yo get the captcha data tho.